rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+There will be a single ROS Service per node, regardless of the number of publishers or subscriptions on that node.
+
+This service may be optional, for example being enabled or disabled when creating the ROS Node.
+
+.. TODO::
+
+    How can we detect when a remote node is not offering this service?
+    It's difficult to differentiate between "the Service has not been created yet, but will be" and "the Service will never be created".
+    Should we use a ROS Parameter to indicate this?
+    But then what if remote access to Parameters (another optional Service) is disabled?
+    Perhaps we need a "services offered" list which is part of the Node metadata, which is sent for each node in the rmw implementation, but that's out of scope for this REP.
+
+A service request to this ROS Service will comprise of the type name and the type version hash, which is distributed during discovery of endpoints and will be accessible through the ROS Graph API, as described in previous sections.
+The ROS Service server will respond with the type description and any necessary metadata needed to do Run-Time Interface Reflection.
+This service is not expected to be called frequently, and is likely to only occur when new topic or service endpoints are created, and even then, only if the endpoint type hashes do not match.
+
+.. TODO::
+
+    Should each endpoint be asked about their type/version pair or should we assume that the type/version pair guarantees a unique type description and therefore reuse past queries?
+    (wjwwood) I am leaning towards assuming the type name and type version hash combo as being a unique identifier.
+
+Type Description Contents and Format
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The response sent by the ROS Service server will contain a combination of the original ``idl`` or ``msg`` file's content, as well as any necessary information to serialize and deserialize the raw message buffers sent on the topic.
+The response will contain a version of the description that contains comments from the original type description, as those might be relevant to interpreting the semantic meaning of the message fields.
+
+Additionally, the response could include the serialization library used, its version, or any other helpful information from the original producer of the data.
+
+.. TODO::
+
+    What happens if the message consumer doesn't have access to the serialization library stated in the meta-type?
+    (wjwwood) It depends on what you're doing with the response. If you are going to subscribe to a remote endpoint, then I think we just refuse to create the subscription, as communication will not work, and there's a basic assumption that if you're going to communicate with an endpoint then you are using the same or compatible rmw implementations, which includes the serialization technology. The purpose of this section and ROS Service is to provide the information, not to ensure communication can happen.
```

Comment:
I think it's fair to say that cross-protocol communications is out of scope for this REP, even if we are providing information that could unlock it down the line.